### PR TITLE
!!! FEATURE: Introduce new fields & refactor import/export

### LINKED ISSUE #18: Add redirect "Type" information
In context of Neos, at least two different providers of redirect entries exist.
1. Automatic generated by node or resource changes in the Neos Backend
2. Manually created via CLI import or the Redirect.Ui

It would be really helpful if these types could be filtered if needed.

### LINKED ISSUE #20: Add and evaluate start & end date time fields for redirects
This would allow preparing redirects for future campaigns and to limit a redirects lifetime.

### LINKED ISSUE #21: Add comment field for redirects
This would allow giving redirects a context why it was added or linking it to a ticket system.

### LINKED ISSUE #22: Add author field for redirects
Storing the author of a redirect as optional text field is helpful in projects in which multiple people work on them.

I wouldn't do this as a reference to an account object as redirects can be imported and exported.

### PR DESCRIPTION
**Pull request by Sebobo**
This changes add several meta fields for redirects:

* comment -> resolves #21 
* creator -> resolves #22 
* startDateTime & endDateTime -> resolves #20 
* type -> resolves #18 

And it also separates the import & export methods from the commands into service classes that can be used in the redirect handler ui.

### CODE REVIEW

File: Classes/Service/RedirectImportService.php [hunk 1]
```diff
@@ -0,0 +1,260 @@
+<?php
+namespace Neos\RedirectHandler\Service;
```
Comment (daniellienert) [inline Classes/Service/RedirectImportService.php]:
```suggestion
declare(strict_types=1);

namespace Neos\RedirectHandler\Service;
```

File: Classes/Service/RedirectImportService.php [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+/*
+ * This file is part of the Neos.Flow package.
```
Comment (daniellienert) [inline Classes/Service/RedirectImportService.php]:
```suggestion
 * This file is part of the Neos.RedirectHandler package.
```

File: Classes/Service/RedirectExportService.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
+<?php
+namespace Neos\RedirectHandler\Service;
+
+/*
+ * This file is part of the Neos.Flow package.
```
Comment (daniellienert) [inline Classes/Service/RedirectExportService.php]:
```suggestion
 * This file is part of the Neos.RedirectHandler package.
```

File: Classes/Command/RedirectCommandController.php [hunk 4]
```diff
@@ -340,7 +297,7 @@ public function removeByHostCommand($host)
      * @param boolean $force Replace existing redirect (based on the source URI)
      * @return void
      */
-    public function addCommand($source, $target, $statusCode, $host = null, $force = false)
+    public function addCommand($source, $target, $statusCode, $host = null, $force = false): void
```
Comment (daniellienert) [inline Classes/Command/RedirectCommandController.php]:
```suggestion
    public function addCommand(string $source, string $target, int $statusCode, ?string $host = null, bool $force = false): void
```

File: Classes/Command/RedirectCommandController.php [hunk 5]
```diff
@@ -377,7 +334,7 @@ public function addCommand($source, $target, $statusCode, $host = null, $force =
      * @param RedirectInterface $redirect
      * @return void
      */
-    protected function outputRedirectLine($prefix, RedirectInterface $redirect)
+    protected function outputRedirectLine($prefix, RedirectInterface $redirect): void
```
Comment (daniellienert) [inline Classes/Command/RedirectCommandController.php]:
```suggestion
    protected function outputRedirectLine(string $prefix, RedirectInterface $redirect): void
```

File: Classes/Exception.php [hunk 6]
```diff
@@ -24,7 +24,8 @@ class Exception extends Http\Exception
      * @param integer $statusCode
      * @return void
      */
-    public function setStatusCode($statusCode) {
+    public function setStatusCode($statusCode)
```
Comment (daniellienert) [inline Classes/Exception.php]:
```suggestion
    public function setStatusCode(int $statusCode)
```

File: Classes/Service/RedirectExportService.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+     * Export redirects to a CSV object than can then be used to write a file or print the content
+     *
+     * @param string $host (optional) Only export hosts for a specified host
+     * @param bool $onlyActive will filter inactive redirects based on start and end datetime if true
+     * @param null|string $type will filter redirects based on their type
+     * @param bool $includeHeader will a header line with column names as first line when true
+     * @return Writer
+     * @throws CannotInsertRecord
+     */
+    public function exportCsv($host = null, $onlyActive = false, $type = null, $includeHeader = true): Writer
```
Comment (daniellienert) [inline Classes/Service/RedirectExportService.php]:
```suggestion
    public function exportCsv(?string $host = null, bool $onlyActive = false, ?string $type = null, bool $includeHeader = true): Writer
```

File: Classes/Redirect.php [hunk 8]
```diff
@@ -47,58 +47,260 @@ class Redirect implements RedirectInterface
...
+     * @param DateTimeInterface|null $endDateTime
+     * @param DateTimeInterface|null $creationDateTime
+     * @param DateTimeInterface|null $lastModificationDateTime
+     * @param int $hitCounter
+     * @param DateTimeInterface|null $lastHit
      */
-    public function __construct($sourceUriPath, $targetUriPath, $statusCode, $host = null)
-    {
+    public function __construct(
+        $sourceUriPath,
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Classes/Redirect.php]: No specific reason, just not used to adding the primitive type hints. Will add them tomorrow.

**Pull request merged**